l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -0,0 +1,316 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+// solhint-disable reason-string, gas-custom-errors
+
+import {Ownable2StepUpgradeable} from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IL1NativeTokenVault} from "./interfaces/IL1NativeTokenVault.sol";
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+import {IL1AssetHandler} from "./interfaces/IL1AssetHandler.sol";
+
+import {IL1SharedBridge} from "./interfaces/IL1SharedBridge.sol";
+import {ETH_TOKEN_ADDRESS, NATIVE_TOKEN_VAULT_VIRTUAL_ADDRESS} from "../common/Config.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev Vault holding L1 native ETH and ERC20 tokens bridged into the hyperchains.
+/// @dev Designed for use with a proxy for upgradability.
+contract L1NativeTokenVault is
+    IL1NativeTokenVault,
+    IL1AssetHandler,
+    ReentrancyGuard,
+    Ownable2StepUpgradeable,
+    PausableUpgradeable
+{
+    using SafeERC20 for IERC20;
+
+    /// @dev The address of the WETH token on L1.
+    address public immutable override L1_WETH_TOKEN;
+
+    /// @dev L1 Shared Bridge smart contract that handles communication with its counterparts on L2s
+    IL1SharedBridge public immutable override L1_SHARED_BRIDGE;
+
+    /// @dev Era's chainID
+    uint256 public immutable ERA_CHAIN_ID;
+
+    /// @dev Indicates whether the hyperbridging is enabled for a given chain.
+    // slither-disable-next-line uninitialized-state
+    mapping(uint256 chainId => bool enabled) public hyperbridgingEnabled;
+
+    /// @dev Maps token balances for each chain to prevent unauthorized spending across hyperchains.
+    /// This serves as a security measure until hyperbridging is implemented.
+    /// NOTE: this function may be removed in the future, don't rely on it!
+    mapping(uint256 chainId => mapping(address l1Token => uint256 balance)) public chainBalance;
+
+    /// @dev Maps token balances for each chain to verify that chain balance was migrated from shared bridge.
+    /// This serves only for migration purposes.
+    mapping(uint256 chainId => mapping(address l1Token => bool migrated)) public chainBalanceMigrated;
```

Comment:
I think a more elegant approach would be to just consume chainbalance from shared bridge, i.e. once it is migrated, just set it to 0.